sites/hashdev/src/_pages/blog/4_announcement-error-stack.mdx

```diff
@@ -0,0 +1,285 @@
+author: "Tim Diekmann"
+jobTitle: "Backend Engineer at HASH"
+date: "2022-06-10"
+---
+
+# **Announcing `error-stack`**
+
+We are proud to announce a new error library for Rust: `error-stack` - a context-aware error library with arbitrary attached user data. **[View the crate on GitHub >](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack)**
+
+This post outlines our trials and tribulations in error-handling, why we ultimately built the crate, what architectural choices we made in doing so, and what you can expect from it.
+
+# Our error-handling journey
+
+At HASH, we've been building a simulation engine — written in Rust — called [hEngine](https://hash.ai/platform/engine). It's a large code base, with a number of contributors, and its size and complexity has grown over time (check out the code on [GitHub](https://github.com/hashintel/hash)). Across some big refactors, as well as everyday feature development, we found error handling became unwieldy and ourselves struggling to keep up. In a bid to improve our debugging process and reduce maintenance overheads we decided to revisit our strategy for managing errors.
+
+Up until recently our approach to error handling centered around using [`thiserror`](https://crates.io/crates/thiserror), a fantastic crate which abstracts away a lot of Rust's verbosity when creating errors. We used to create enumerations (generally a large, top-level one, per crate and/or major modules) to describe our different error types, and used `thiserror`'s macros to handle a lot of the trait implementations. We didn't want to be using `Box<dyn Error>` everywhere, meaning we wanted typed errors, so we utilized `thiserror`'s ability to generate `From` implementations from a macro to handle the transformation of our error types across the module and crate boundaries. (`thiserror` lets you succinctly define a `From` implementation like so: `IoError(#[from] std::io::Error)`).
+
+## Handling Errors across Module and Crate Boundaries
+
+This led us into a pattern where we'd have a function called across a boundary (say from the `engine` module to the `simulation` crate), necessitating that we transform it between error types, and so for ergonomic reasons, we'd create a `From` implementation for the whole `Error` enumeration:
+
+```rust
+// mod engine
+
+#[derive(ThisError)]
+enum Error {
+    #[error("Simulation error: {0}")]
+    Simulation(#[from] crate::simulation::Error),
+    ...
+}
+```
+
+This would then often turn into error messages that looked like the following:
+
+```
+Error: Engine Error: Simulation Error: Experiment Error: Datastore Error: No such file or directory (os error 2)
+```
+
+These error messages would tell us very little about the context that led to the error and weren't very helpful in actually tracking down the causes of problems. And while we have logging infrastructure in place, which helps with tracking down events that lead to errors, it relies on a few things:
```

Comment:
> helpful in [...] tracking down

> helps with tracking down

Can we come up with a better phrasing?

Reply:
This phrasing is good imo, is there something specific you don't like about it?

I'm also good with "which is helpful in tracking down events that lead to errors"

Reply:
It's mainly because of using the same phrasing twice directly next to each other. I don't have a strong opinion, only wanted feedback from a native English speaker 🙂

Reply:
Ah sorry I didn't actually spot the repeated phrase! Good call-out, although it's not too egregious here haha.

```suggestion
These error messages would tell us very little about the context that led to the error and weren't very helpful in actually tracking down the causes of problems. And while we have logging infrastructure in place, which is helpful with identifying events that lead to errors, it relies on a few things:
```

Reply:
Sounds good, thank you!